File: 0692-top-k-frequent-words/0692-top-k-frequent-words.cs
public class Solution {
    public IList<string> TopKFrequent(string[] words, int k) {
        int n = words.Length;
        var result = new List<string>();

        // 1. Count frequencies
        var count = new Dictionary<string, int>();
        foreach (var w in words) {
            if (!count.ContainsKey(w)) {
                count[w] = 0;
            }
            count[w]++;
        }

        // 2. Buckets: index = frequency, each bucket holds words with that freq
        var buckets = new List<string>[n + 1];
        foreach (var kvp in count) {
            string word = kvp.Key;
            int freq = kvp.Value;
            if (buckets[freq] == null) {
                buckets[freq] = new List<string>();
            }
            buckets[freq].Add(word);
        }

        // 3. Walk buckets from high freq to low
        for (int freq = n; freq > 0 && result.Count < k; freq--) {
            if (buckets[freq] == null) continue;

            // For equal frequency, words must be in lexicographic order
            buckets[freq].Sort(StringComparer.Ordinal);

            foreach (var word in buckets[freq]) {
                result.Add(word);
                if (result.Count == k) {
                    return result;
                }
            }
        }

        return result;
    }
}

File: 0692-top-k-frequent-words/0692-top-k-frequent-words.java
class Solution {
    public List<String> topKFrequent(String[] words, int k) {
                Map<String, Integer> freq = new HashMap<>();
        for (String w : words) {
            freq.put(w, freq.getOrDefault(w, 0) + 1);
        }

        List<String> candidates = new ArrayList<>(freq.keySet());

        // Sort by: frequency desc, then word asc
        Collections.sort(candidates, (w1, w2) -> {
            int f1 = freq.get(w1);
            int f2 = freq.get(w2);
            if (f1 != f2) {
                return Integer.compare(f2, f1); // higher freq first
            }
            return w1.compareTo(w2);            // lexicographically smaller first
        });

        return candidates.subList(0, k);
    }
}